Extract AggregationFuzzerBase from AggregationFuzzer

kagamiori opened the pull request:
Summary: Extract AggregationFuzzerBase from AggregationFuzzer. This is needed for building WindowFuzzer
that reuses common logic in AggregaitonFuzzerBase.

This diff also moves fuzzer-related files to 
velox/exec/fuzzer, including
– AggregationFuzzer.h/cpp
– AggregationFuzzerRunner.h
– ReferenceQueryRunner.h
– DuckQueryRunner.h/cpp
– PrestoQueryRunner.h/cpp
– InputGenerator.h
– ResultVerifier.h
– TransformResultVerifier.h

It also moves the Presto functions' custom input generators and result verifiers to 
velox/functions/prestosql/fuzzer, including
– MinMaxInputGenerator.h
– ApproxDistinctInputGenerator.h
– ApproxDistinctResultVerifier.h
– ApproxPercentileInputGenerator.h
– ApproxPercentileResultVerifier.h

This is the first piece of https://github.com/facebookincubator/velox/issues/7754.

Differential Revision: D51692940

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzerBase.h:
```diff
@@ -0,0 +1,344 @@
+
+constexpr const std::string_view kPlanNodeFileName = "plan_nodes";
+
+class InputGenerator {
```
Let's move this into its own file. Same for ResultVerifier.

kagamiori commented:
Hi @mbasmanova, thanks for the suggestions! I have moved those classes to their own files. Please let me know if you have any other thoughts.

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzerUtils.h:
```diff
@@ -0,0 +1,98 @@
+// Parse the comma separated list of function names, and use it to filter the
+// input signatures.
+template <typename SignatureMapType>
+SignatureMapType filterSignatures(
```
There is similar (if not identical) function in velox/expression/tests/ExpressionFuzzer.cpp. Perhaps, move it to FuzzerToolkit.h and reuse in both places.

mbasmanova replied:
@kagamiori I see that you moved filterSignatures into FuzzerToolkit.h, but I don't see changes to ExpressionFuzzer.cpp to use it. An I missing something?

kagamiori replied:
Hi @mbasmanova, the filterSignatures() function used in ExpressionFuzzer is a bit different from the one moved to FuzzerToolkit: (1) ExpressionFuzzer's filterSignature() modifies the input signature map in place rather than returning a new object, and (2) ExpressionFuzzer's filterSignature() handles function signatures in addition to function names in the skip function list. So it's not a simple replacement. Could we make this change in the separate diff since the current one is already too large?

mbasmanova replied:
Sure. We can consolidate these in a follow-up. Let's add a TODO for now. Thanks.

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzerUtils.h:
```diff
@@ -0,0 +1,98 @@
+// Parse the comma separated list of function names and return a set of them.
+std::unordered_set<std::string> splitNames(const std::string& names);
+
+// Parse the comma separated list of function names, and use it to filter the
```
This comment needs updating.

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzerUtils.h:
```diff
@@ -0,0 +1,98 @@
+
+// Parse the comma separated list of function names, and use it to filter the
+// input signatures.
+template <typename SignatureMapType>
```
Why is this a template?

kagamiori replied:
It is templated on the input and output type so that it can process both AggregateFunctionSignatureMap and WindowFunctionMap.

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzerUtils.h:
```diff
@@ -0,0 +1,98 @@
+};
+
+// Parse the comma separated list of function names and return a set of them.
+std::unordered_set<std::string> splitNames(const std::string& names);
```
Let's put this into "detail" namespace as this is not part of "public" API.

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzerUtils.h:
```diff
@@ -0,0 +1,98 @@
+
+namespace facebook::velox::exec::test {
+
+struct AggregationFuzzerOptions {
```
Once filterSignatures are moved into FuzzerToolkit, this will be the only thing left here. We can then rename the file to AggregationFuzzerOptions.h

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzer.cpp:
```diff
@@ -177,66 +116,8 @@ class AggregationFuzzer {
-      std::vector<TypePtr> types,
-      const CallableSignature& signature);
+  void updateReferenceQueryStats(
+      AggregationFuzzerBase::ReferenceQueryErrorCode ec);
```
ec -> errorCode

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzer.h:
```diff
@@ -16,82 +16,12 @@
 #pragma once
 
 #include "velox/exec/Aggregate.h"
+#include "velox/exec/tests/utils/AggregationFuzzerBase.h"
```
I wonder if it makes sense to create exec/fuzzer subdirectory and move fuzzer-related code there. exec/tests could still have AggregationFuzzerTest and AggregationSparkFuzzerTest. And also move input generators and custom verifiers to something liek functions/prestosql/fuzzer. In fact, maybe we could setup something like this:

exec/fuzzer - common fuzzer classes
functions/prestosql/fuzzer - AggregationFuzzerTest + input generators and custom verifiers for Presto functions
functions/sparksql/fuzzer - SparkAggregationFuzzerTest + input generators and custom verifiers for Spark functions

mbasmanova reviewed (commented):
@kagamiori Overall looks good. Some follow-up comments.

kagamiori commented:
Hi @mbasmanova, I've moved fuzzer-related files to velox/exec/fuzzer and the input generator files to velox/functions/prestosql/fuzzer. This PR doesn't move the AggregationFuzzerTest.cpp and SparkAggregationFuzzerTest.cpp files. I move them in https://github.com/facebookincubator/velox/pull/8007. I split them into a separate PR because the current PR is already too large, and moving the test targets may require additional changes to our configuration files of the PR-time and nightly tests, so I'd like to do that separately from the current PR.

mbasmanova commented on velox/exec/tests/AggregationFuzzerTest.cpp:
```diff
@@ -20,15 +20,17 @@
-#include "velox/exec/tests/utils/DuckQueryRunner.h"
 #include "velox/exec/tests/utils/PlanBuilder.h"
 #include "velox/functions/prestosql/aggregates/RegisterAggregateFunctions.h"
+#include "velox/functions/prestosql/fuzzer/ApproxDistinctInputGenerator.h"
```
Thank you for organizing code nicely. Would you update PR description to explain how the new code layout?

mbasmanova commented on velox/exec/tests/AggregationFuzzerTest.cpp:
```diff
@@ -20,15 +20,17 @@
 #include "velox/exec/tests/utils/PlanBuilder.h"
 #include "velox/functions/prestosql/aggregates/RegisterAggregateFunctions.h"
+#include "velox/functions/prestosql/fuzzer/ApproxDistinctInputGenerator.h"
+#include "velox/functions/prestosql/fuzzer/ApproxPercentileInputGenerator.h"
```
I see that you moved InputGenerators, but not ResultVerifiers. Is there any particular reason to keep verifiers here?

mbasmanova commented on velox/expression/tests/utils/FuzzerToolkit.h:
```diff
@@ -57,4 +57,46 @@ void sortSignatureTemplates(std::vector<SignatureTemplate>& signatures);
+// input signatures. Return a signature map that (1) only include functions
+// appearing in onlyFunctions if onlyFunctions is non-empty, and (2) not include
+// any functions appearing in skipFunctions if skipFunctions is non-empty.
+template <typename SignatureMapType>
```
Would you add @tparam comment to explain what are acceptable types for SignatureMapType ?

mbasmanova reviewed (commented):
@kagamiori Wei, thank you for iterating on this PR. Overall looks good % some questions and CI being all red.

kagamiori commented:
> @kagamiori Wei, thank you for iterating on this PR. Overall looks good % some questions and CI being all red.

Hi @mbasmanova, thank you for the review! I've addressed your comments. I've tested the build on Mac using cmake, so the red signals should be gone this time. Could you take a look again? Thanks!

mbasmanova reviewed (approved):
Thanks.
